modules/monitoring/includes/downtime.php

```diff
@@ -0,0 +1,444 @@
+	 */
+	public function get_scheduled_end() {
+		$date_str = $this->model->get_end_date();
+		$time_str = $this->model->get_end_time_string();
+		return $this->dt_from_str($date_str, $time_str);
+	}
+}
+
+/**
+ * RecurringDowntime class
+ *
+ * Used with recurring schedules, inherits from Downtime.
+ */
+class RecurringDowntime extends Downtime {
+	/**
+	 * object recurrence object
+	 */
+	public $recurrence;
+
+	/**
+	 * array nested recurrence map
+	 */
+	public $recurrence_on;
+
+	/**
+	 * NinjaDateTime
+	 */
+	public $recurrence_ends;
+
+	/**
+	 * array
+	 */
+	public $exclude_days;
+
+	/**
+	 * Provides a helpers for working with recurring downtime.
+	 *
+	 * @param $model object
+	 * @throws Exception UnexpectedValueException
+	 * @throws NoRecurrenceException
+	 */
+	function __construct($model) {
+		parent::__construct($model);
+
+		$this->recurrence = $this->get_recurrence();
+		$this->recurrence_on = $this->get_recurrence_on();
+		$this->recurrence_ends = $this->get_recurrence_ends();
+		$this->exclude_days = explode(',', $model->get_exclude_days());
+	}
+
+	/**
+	 * Returns recurrence for schedule.
+	 *
+	 * @return object
+	 * @throws Exception UnexpectedValueException
+	 * @throws NoRecurrenceException
+	 */
+	private function get_recurrence() {
+		$recurrence = json_decode($this->model->get_recurrence());
+		if(!$recurrence) {
+			throw new NoRecurrenceException('RecurringSchedule.schedule must have a recurrence');
+		}
+
+		return $recurrence;
+	}
+
+	/**
+	 * Returns recurrence for schedule.
+	 *
+	 * @return array decoded recurrence_on
+	 */
+	private function get_recurrence_on() {
+		return json_decode($this->model->get_recurrence_on(), true);
+	}
+
+	/**
+	 * Returns recurrence end date as a NinjaDateTime object.
+	 *
+	 * @return NinjaDateTime
+	 * @throws Exception
+	 */
+	private function get_recurrence_ends() {
+		$ends = $this->model->get_recurrence_ends();
+		if(!$ends) {
+			// Return never-ending recurrences as a NinjaDateTime object as well, for consistency.
+			return new NinjaDateTime('2199-12-31');
+		}
+		return new NinjaDateTime($this->model->get_recurrence_ends());
+	}
+
+	/**
+	 * Checks if the given $target_date is excluded.
+	 *
+	 * The input, $target_date, is a NinjaDateTime object or--to simplify testing--a
+	 * parsable date string, that is compared against an array of ranges or single
+	 * dates contained in `$this->exclude_days`:
+	 * ['2019-03-13 to 2019-03-16', '2019-03-13 to 2019-03-18', ...]
+	 *
+	 * @param $target_date mixed NinjaDateTime|str
+	 * @return bool
+	 * @throws Exception UnexpectedValueException Invalid date-exclusion format
+	 */
+	public function is_excluded($target_date) {
+		if(gettype($target_date) === 'string') {
+			$target_date = new NinjaDateTime($target_date);
+		}
+
+		foreach($this->exclude_days as $item) {
+			$parsed = explode('to', $item);
+			if(sizeof($parsed) === 1) {
+				// Single date, set to and from to idx 0.
+				$from = $to = trim($parsed[0]);
```

Comment:
Just curious about how does this work. Is it some kind of an if statement?

Reply:
It sets both `$from` and `$to` to be equal `trim($parsed[0])` :-)

Reply:
Oh cool! 😮